[dnf5] Check packages signatures before installation

m-blaha opened the pull request.

m-blaha committed:
libdnf/rpm: Create new RpmLogGuardStrings

This class acquires rpm log mutex and then redirects all rpm logs into a
vector of strings from which they could be later read.
Needed because rpm API rpmcliVerifySignatures used for signature
checking does not return all needed information as a result, but to get
details about a failure the rpm logs have to be parsed.

m-blaha committed:
libdnf/rpm: Created class for handling rpm package signatures

The new RpmSignature class now can verify signature of given rpm
package.

m-blaha committed:
common/url: Share is_url() between libdnf and dnf5

m-blaha committed:
libdnf/repo: Share rawkey2infos() accross libdnf

The function will be needed also for importing keys into rpm database.

m-blaha committed:
dnf5: check packages signatures before installation

jrohel commented on libdnf/repo/repo_gpgme.hpp:
```diff
@@ -25,13 +25,38 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "libdnf/repo/config_repo.hpp"
 #include "libdnf/repo/repo_callbacks.hpp"
 
+#include <gpgme.h>
```
Because #include gpgme.h is moved here, please remove it from "repo_gpgme.cpp".

jrohel commented on libdnf/rpm/rpm_log_guard.cpp:
```diff
@@ -76,4 +75,18 @@ RpmLogGuard::~RpmLogGuard() {
 
+static int rpmlog_callback_strings(rpmlogRec rec, rpmlogCallbackData data) {
+    std::string msg(rpmlogRecMessage(rec));
+    if (!msg.empty() and msg[msg.length() - 1] == '\n') {
```
Yes C++ supports the `and` operator, but I prefer to use one type of syntax (`&&`) in a project.

m-blaha replied:
Sorry, that's my Python background seeping through. Fixed.

jrohel commented on include/libdnf/rpm/rpm_signature.hpp:
```diff
@@ -0,0 +1,110 @@
+#include "libdnf/common/exception.hpp"
+#include "libdnf/rpm/package.hpp"
+
+#include <rpm/rpmts.h>
```
Please remove this include from the public header file.

Example solution:
* Move the line `using RpmTransactionPtr = std::unique_ptr<rpmts_s, std::function<void(rpmts_s * ts)>>;` to the .cpp file.
* Remove private methods `RpmTransactionPtr create_transaction() const;` and 
`bool rpmdb_lookup(const RpmTransactionPtr & ts_ptr, const KeyInfo & key) const;` from RpmSignature class.
* and change them to static functions in .cpp:
`static RpmTransactionPtr create_transaction(const BaseWeakPtr & base) {` and 
`static bool rpmdb_lookup(const RpmTransactionPtr & ts_ptr, const KeyInfo & key) {`
* and call function `create_transaction` with the `base` parameter.

jrohel commented on libdnf/rpm/rpm_signature.cpp:
```diff
@@ -0,0 +1,210 @@
+        key_id = key_info.get_id();
+        user_id = key_info.get_user_id();
+        fingerprint = key_info.get_fingerprint();
+    }
```
What is the purpose of this loop?
`rawkey2infos` returns a vector with information about signing subkeys. The loop iterates through the vector and the next key information overwrites the previous one. So at the end we get information about the last signing subkey. Why is only the last subkey used? And why don't we  get the last subkey directly (the key from position size() - 1) instead of  looping through all the items in the vector?

m-blaha replied:
As I understand it, `rawkeys2infos` always extracts only the first signing subkey.
But I see the issue now - to reproduce it you need a file with multiple public keys. Having a key with multiple subkeys is not enough.
With such key file dnf4 will ask for confirmation of all the keys, current dnf5 implementation imports only the first one.
I'll fix it.

m-blaha replied:
Multiple keys issue should be fixed by the latest commit.

m-blaha committed:
libdnf: Fix reading GPG key from file

m-blaha committed:
libdnf/rpm: Allow multiple GPG keys in public key file

In case the repository public key file contains more keys, they are
imported and user-confirmed independently.

jrohel commented on dnf5/context.cpp:
```diff
@@ -667,9 +661,139 @@ std::chrono::time_point<std::chrono::steady_clock> RpmTransCB::prev_print_time =
 }  // namespace
 
+
+bool user_confirm_key(libdnf::ConfigMain & config, const libdnf::rpm::KeyInfo & key_info) {
```
The function is not listed in the header file. Is it only used in this cpp file? So make it part of anonymous namespace (ended just before this function ) or static.

m-blaha replied:
Done.

m-blaha committed:
fixup! dnf5: check packages signatures before installation

jrohel commented:
Thanks.
